### PR TITLE
Langfuse support

### PR DESCRIPTION
**Pull request by kongzii**
Hi, as promised in https://github.com/galatolofederico/microchain/issues/16, support for Langfuse.

If you would like to take a look, I made one of our runs publicly available at https://cloud.langfuse.com/project/clv2hvvyw0006z9uchz6u1irw/traces/cf235317-5582-4848-bf29-84625b0f8f7d?observation=df817f99-a127-4cc6-9f0c-eb950b193539:

Screenshot by Dropbox Capture

### CODE REVIEW

File: microchain/engine/agent.py [hunk 1]
```diff
@@ -24,10 +24,28 @@ def __init__(self, llm, engine, on_iteration_start=None, on_iteration_step=None,
...
         self.stop_list = stop_list
+        self.enable_langfuse = enable_langfuse
 
         self.engine.bind(self)
         self.reset()
 
+        if self.enable_langfuse:
+            self.init_langfuse()
+
+    def init_langfuse(self):
```
Comment (evangriffiths) [inline microchain/engine/agent.py:35]: What do you think about calling `engine.init_langfuse()` here (which in turn also calls `generator.init_langfuse()` and `function.init_langfuse()` for each registered function)? Then the only the `Agent` class would have a `enable_langfuse: bool` flag.

Comment (kongzii) [reply microchain/engine/agent.py:35]: Implemented partially, generator classes (ie `OpenAITextGenerator`) now do not accept `enable_langfuse` and instead, they are initialized inside of the agent:

Screenshot by Dropbox Capture

However, then the issue is, that engine class doesn't have `enable_langfuse` argument (it wasn't needed) and If I understand correctly you don't want to introduce that parameter there either, but if we take a look at our PMA codebase:

Screenshot by Dropbox Capture

So the function still need to call `self.init_langfuse()` inside of its `bind` method.

Or maybe you had some other idea around it?

Comment (evangriffiths) [reply microchain/engine/agent.py:35]: Ohh I hadn't thought about that. Either way, I think doing `function.init_langfuse()` on binding makes sense, so I'm happy with your solution!

**Comment by galatolofederico** [comment]
Hi! thanks a lot for this PR! I reviewed the code and everything seems ok. I have not tested it, i will merge and create a release and i will test the integration as soon as possible, thanks again!

**Comment by galatolofederico** [comment]
I’ve added commit 44ef55a5b69495089867864cb7550ca7facd29fe , which introduces the `enable_langfuse` argument to the generators and stores it as an attribute for subsequent checks. Could you please test it with Langfuse and let me know if everything works as expected?

**Comment by kongzii** [comment]
>  Could you please test it with Langfuse and let me know if everything works as expected?

After the fix in the other PR, all looks good so far! 🙏

**Comment by galatolofederico** [comment]
Perfect thank you again!

**Pull request merged**